File: t/PACSFramework/CS/MedicalWorkstationCustomizationDemo/Examples/CustomizeActionsExample/ActionsEditorViewPresenter.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Leadtools.Medical.Workstation.Presenters;
using Leadtools.Medical.Workstation;
using Leadtools.MedicalViewer;
using System.Windows.Forms;
using Leadtools.Medical.Workstation.UI;
using System.ComponentModel;

namespace Leadtools.Demos.Workstation.Customized
{
   class ActionsEditorViewPresenter : WorkstationPresenterBase <IActionsEditorView>
   {
      #region Public
         
         #region Methods
         
         #endregion
         
         #region Properties
            
         #endregion
         
         #region Events
            
         #endregion
         
         #region Data Types Definition
            
         #endregion
         
         #region Callbacks
            
         #endregion
         
      #endregion
      
      #region Protected
         
         #region Methods
         
            protected override void DoInitialize 
            (
               WorkstationContainer container, 
               IActionsEditorView view
            )
            {
               MedicalViewerActionType [] actions ;
               
               
               if ( container.State.DataServices.IsRegistered <ActionsEditorState> ( ) )
               {
                  __ActionsEditorState = container.State.DataServices.Get <ActionsEditorState> ( ) ;
               }
               else
               {
                  __ActionsEditorState = new ActionsEditorState ( ) ;
                  
                  container.State.DataServices.Register <ActionsEditorState> ( __ActionsEditorState ) ;
               }
               
               actions = GetAllowedActions ( container ) ;
               
               view.SetActions ( actions ) ;
               view.SetActionMouseButtons ( Enum.GetValues ( typeof ( MedicalViewerMouseButtons ) ).OfType <MedicalViewerMouseButtons> ( ).ToArray ( ) ) ;
               
               if ( actions.Length > 0 )
               {
                  view.SelectedAction = actions [ 0 ] ;
                  
                  SetActionInformation ( view, container, actions [ 0 ] ) ;
                  
                  RegisterViewEvents ( view ) ;
               }
               else
               {
                  view.CanAddAction    = false ;
                  view.CanRemoveAction = false ;
               }
               
               view.ActivateView ( container.State.ActiveWorkstation ) ;
            }

         #endregion
         
         #region Properties
            
         #endregion
         
         #region Events
            
         #endregion
         
         #region Data Members
            
         #endregion
         
         #region Data Types Definition
            
         #endregion
         
      #endregion
      
      #region Private
         
         #region Methods
         
            private MedicalViewerActionType[] GetAllowedActions ( WorkstationContainer container )
            {
               List <MedicalViewerActionType> actions ;
               
               
               actions = new List<MedicalViewerActionType> ( ) ;
               
               foreach ( MedicalViewerActionType action in Enum.GetValues ( typeof ( MedicalViewerActionType ) ).OfType <MedicalViewerActionType> ( ).ToArray ( )  ) 
               {
                  if ( action == MedicalViewerActionType.None || 
                       action == MedicalViewerActionType.Rotate3DCamera ||
                       action == MedicalViewerActionType.Rotate3DObject ||
                       action == MedicalViewerActionType.RotatePlane ||
                       action == MedicalViewerActionType.Scale3DObject ||
                       action == MedicalViewerActionType.Translate3DCamera ||
                       action == MedicalViewerActionType.TranslatePlane  )
                  {
                     continue ;
                  }
                  
                  if ( container.State.MedicalViewerCellActions.Contains ( action ) )
                  {
                     if ( __ActionsEditorState.CustomActions.Contains ( action ) )
                     {
                        actions.Add ( action ) ;
                     }
                  }
                  else
                  {
                     actions.Add ( action ) ;
                  }
               }
               
               return actions.ToArray ( ) ;
            }
            
            
            private void SetActionInformation
            (
               IActionsEditorView view, 
               WorkstationContainer container,
               MedicalViewerActionType action
            )
            {
               bool isNewAction ;
               
               
               isNewAction = !__ActionsEditorState.CustomActions.Contains ( action ) ;
               
               if ( isNewAction )
               {
                  view.CanAddAction    = true ;
                  view.CanRemoveAction = false ;
                  
                  view.ActionDisplayName   = action.ToString ( ) ;
                  view.SelectedMouseButton = MedicalViewerMouseButtons.Left ;
                  
                  view.CanEditToolstripButtons = true ;
                  
                  view.FeatureId = action.ToString ( ) + "FeatureId" ;
                  
                  view.ToolStipItemImage            = null ;
                  View.ToolStipItemAlternativeImage = null ;
               }
               else
               {
                  view.CanAddAction    = false ;
                  view.CanRemoveAction = true ;
                  
                  var keyValuePair = container.State.MedicalViewerCellMouseButtonActions.Where ( n=>n.Value == action ).FirstOrDefault ( ) ;
                  
                  view.ActionDisplayName   = container.State.WorkstationActionDisplayName [ action ] ;
                  view.SelectedMouseButton = keyValuePair.Key ;
                  
                  view.CanEditToolstripButtons = keyValuePair.Key == MedicalViewerMouseButtons.Left ;
                  
                  if ( __ActionsEditorState.ActionAssociatedFeature.ContainsKey ( action ) )
                  {
                     view.FeatureId = __ActionsEditorState.ActionAssociatedFeature [ action ] ;
                     
                     if ( !string.IsNullOrEmpty ( view.FeatureId ) )
                     {
                        ToolStripItem[] items = container.StripItemFeatureExecuter.GetItems ( view.FeatureId ) ;
                        
                        if ( items.Length > 0 && items [ 0 ] is IToolStripItemItemProperties )
                        {
                           view.ToolStipItemImage            = ( ( IToolStripItemItemProperties ) items [ 0 ] ).ItemProperties.DefaultImage ;
                           view.ToolStipItemAlternativeImage = ( ( IToolStripItemItemProperties ) items [ 0 ] ).ItemProperties.AlternativeImage ;
                        }
                     }
                  }
               }
               
               view.CanChangeDisplayName = view.CanAddAction ;
               view.CanChangeFeatureId   = view.CanAddAction ;
               view.CanChangeMouseButton = view.CanAddAction ;
            }
            
            private void RegisterViewEvents ( IActionsEditorView view )
            {
               view.SelectedActionChanged            += new EventHandler ( view_SelectedActionChanged ) ;
               view.SelectedActionMouseButtonChanged += new EventHandler ( view_SelectedActionMouseButtonChanged ) ;
               view.AddSelectedAction                += new EventHandler ( view_AddSelectedAction ) ;
               view.RemoveSelectedAction             += new EventHandler ( view_RemoveSelectedAction ) ;
               view.ValidateFeatureId += new System.ComponentModel.CancelEventHandler(view_ValidateFeatureId);
            }

            void view_ValidateFeatureId ( object sender, CancelEventArgs e )
            {
               e.Cancel = false ;
               View.SetValidationErrorText ( string.Empty ) ;
               
               if ( string.IsNullOrEmpty ( View.FeatureId ) )
               {
                  e.Cancel = true ;
                  
                  View.SetValidationErrorText ( "Feature ID can't be empty" ) ;
               }
               else
               {
                  if ( ViewerContainer.FeaturesFactory.IsCommandRegistered ( View.FeatureId ) )
                  {
                     e.Cancel = true ;
                     
                     View.SetValidationErrorText ( "Feature ID already registered" ) ;
                  }
               }
            }

            private void AddCustomMenuItem
            (
               MedicalViewerActionType action, 
               WorkstationMenuProperties newMenuProperties, 
               ToolStripDropDownItem item
            )
            {
               CustomToolStripMenuItem newActionItem ;
               
               
               
               newActionItem = new CustomToolStripMenuItem ( newMenuProperties ) ;
               
               
               item.DropDownItems.Insert ( 0, newActionItem ) ;
               
               ViewerContainer.State.ActiveWorkstation.AddToolbarSplitButtonChildItemLeftAction ( (CustomToolStripSplitButton) item,
                                                                                                   newActionItem,
                                                                                                   action ) ;
                                                                                                   
               ViewerContainer.StripItemFeatureExecuter.SetItemFeature ( newActionItem ) ;
            }

            private void AddCustomButton
            ( 
               MedicalViewerActionType action, 
               WorkstationMenuProperties newMenuProperties, 
               ToolStripDropDownItem item 
            )
            {
               CustomToolStripButton newActionItem ;
               
               
               newActionItem = new CustomToolStripButton ( newMenuProperties ) ;
               
               item.DropDownItems.Add ( newActionItem ) ;
               
               ViewerContainer.State.ActiveWorkstation.AddToolbarActivatedButtonLeftAction ( item, action ) ;
               
               ViewerContainer.StripItemFeatureExecuter.SetItemFeature ( newActionItem ) ;
            }
            
            private void AddDesignerItem
            ( 
               MedicalViewerActionType action, 
               WorkstationMenuProperties newMenuProperties, 
               ToolStripDropDownItem item 
            )
            {
               DesignToolStripMenuItem newActionItem ;
               
               
               newActionItem = new DesignToolStripMenuItem ( newMenuProperties ) ;
               
               item.DropDownItems.Add ( newActionItem ) ;
               
               ViewerContainer.StripItemFeatureExecuter.SetItemFeature ( newActionItem ) ;
            }

            private void AddActionToCurrentCells ( MedicalViewerActionType action )
            {
               foreach ( StudiesViewer studyViewer in ViewerContainer.State.WorkstationHostViewers.OfType <StudiesViewer> ( ) )
               {
                  foreach ( MedicalViewer.MedicalViewer medicalViewer in studyViewer.Viewers ) 
                  {
                     foreach ( MedicalViewerCell cell in medicalViewer.Cells )
                     {
                        cell.AddAction ( action ) ;
                     }
                  }
               }
            }

            private ToolStripItem[] GetAnnotationOrRegionsParnetItems ( MedicalViewerActionType action )
            {
               if ( action.ToString ( ).Contains ( "Annotation" ) )
               {
                  return ToolStripMenuProperties.Instance.AnnotationstoolStripSplitButton.AssociatedItems.OfType <ToolStripDropDownItem> ( ).ToArray ( ) ;
               }
               else if ( action.ToString ( ).Contains ( "Region" ) )
               {
                  return ToolStripMenuProperties.Instance.RegionstoolStripSplitButton.AssociatedItems.OfType <ToolStripDropDownItem> ( ).ToArray ( ) ;
               }
               else
               {
                  return null ;
               }
            }
            
         #endregion
         
         #region Properties
         
            private ActionsEditorState __ActionsEditorState
            {
               get ;
               set ;
            }
            
         #endregion
         
         #region Events
         
            void view_AddSelectedAction ( object sender, EventArgs e )
            {
               MedicalViewerActionType action = View.SelectedAction ;
               ApplyActionCommand actionCommand ;
               
               if ( !ViewerContainer.State.MedicalViewerCellActions.Contains ( action ) )
               {
                  AddActionToCurrentCells ( action ) ;
                  
                  ViewerContainer.State.MedicalViewerCellActions.Add ( action ) ;
               }
               
               ViewerContainer.State.WorkstationActionDisplayName [action] = View.ActionDisplayName ;
               
               __ActionsEditorState.CustomActions.Add ( action ) ;
               
               actionCommand = new ApplyActionCommand ( View.FeatureId, ViewerContainer, action, View.SelectedMouseButton, ApplyActionFlags.Viewer2D ) ;
               
               if ( View.SelectedMouseButton == MedicalViewerMouseButtons.Left )
               {
                  WorkstationMenuProperties newMenuProperties ;
                  
                  
                  newMenuProperties           = new WorkstationMenuProperties ( ViewerContainer.State.WorkstationActionDisplayName [action], View.ToolStipItemImage, View.ToolStipItemAlternativeImage ) ;
                  newMenuProperties.FeatureId = View.FeatureId ;
                  
                  __ActionsEditorState.ActionAssociatedFeature [ action ] = View.FeatureId ;
                  
                  ViewerContainer.FeaturesFactory.RegisterCommand ( actionCommand ) ;
                  
                  ToolStripItem[] parentItems = GetAnnotationOrRegionsParnetItems ( action ) ;
                  
                  if(  null != parentItems )
                  {
                     foreach ( ToolStripDropDownItem item in parentItems ) 
                     {
                        if ( ViewerContainer.State.ActiveWorkstation.IsToolbarActivatedButtonLeftAction ( item ) )
                        {
                           if ( item is CustomToolStripSplitButton ) 
                           {
                              AddCustomMenuItem ( action, newMenuProperties, item ) ;
                           }
                           else
                           {
                              AddCustomButton(action, newMenuProperties, item);
                           }
                        }
                        else
                        {
                           AddDesignerItem ( action, newMenuProperties, item ) ;
                        }
                     }
                  }
                  else
                  {
                     CustomToolStripSplitButton annotationToolbarItem = null ;
                     CustomToolStripButton newActionItem ;
                              
                              
                              
                     newActionItem = new CustomToolStripButton ( newMenuProperties ) ;
                     
                     foreach ( CustomToolStripSplitButton annotationItem in ToolStripMenuProperties.Instance.AnnotationstoolStripSplitButton.AssociatedItems.OfType <CustomToolStripSplitButton> ( ) )
                     {
                        if ( annotationItem.Owner == ViewerContainer.State.ActiveWorkstation.ViewerToolbar )
                        {
                           annotationToolbarItem = annotationItem ;
                           
                           break ;
                        }
                     }
                     
                     if ( null != annotationToolbarItem )
                     {
                        ViewerContainer.State.ActiveWorkstation.ViewerToolbar.Items.Insert ( ViewerContainer.State.ActiveWorkstation.ViewerToolbar.Items.IndexOf ( annotationToolbarItem ) - 1, 
                                                                                             newActionItem ) ;
                     }
                     else
                     {
                        ViewerContainer.State.ActiveWorkstation.ViewerToolbar.Items.Add ( newActionItem ) ;
                     }
                     
                     ViewerContainer.State.ActiveWorkstation.AddToolbarActivatedButtonLeftAction ( newActionItem, action ) ;
                     
                     ViewerContainer.StripItemFeatureExecuter.SetItemFeature ( newActionItem ) ;
                  }
               }
             
               if ( actionCommand.CanExecute ( ) )
               {
                  actionCommand.Execute ( ) ;
               }
               
               SetActionInformation ( View, ViewerContainer, View.SelectedAction ) ;
            
            }
            
            void view_RemoveSelectedAction ( object sender, EventArgs e )
            {
               MedicalViewerActionType action = View.SelectedAction ;
               
               
               if ( __ActionsEditorState.CustomActions.Contains ( action ) )
               {
                  if ( ViewerContainer.State.MedicalViewerCellActions.Contains ( action ) )
                  {
                     ViewerContainer.State.MedicalViewerCellActions.Remove ( action ) ;
                  }
                  
                  ViewerContainer.State.WorkstationActionDisplayName.Remove ( action ) ;
                  
                  __ActionsEditorState.CustomActions.Remove ( action ) ;
               }
               
               if ( __ActionsEditorState.ActionAssociatedFeature.ContainsKey ( action ) )
               {
                  ToolStripItem [] items ;
                  
                  
                  items = ViewerContainer.StripItemFeatureExecuter.GetItems ( __ActionsEditorState.ActionAssociatedFeature [ action ]) ;
                  
                  foreach ( ToolStripItem item in items ) 
                  {
                     if ( item is ActivatedButton && ViewerContainer.State.ActiveWorkstation.IsToolbarActivatedButtonLeftAction ( item ) )
                     {
                        ViewerContainer.State.ActiveWorkstation.RemoveToolbarActivatedButtonLeftAction ( item, action ) ;
                     }
                     else if ( item.OwnerItem is ActivatedButton )
                     {
                        ViewerContainer.State.ActiveWorkstation.RemoveToolbarActivatedButtonLeftAction ( item.OwnerItem, action ) ;
                     }
                     
                     ViewerContainer.StripItemFeatureExecuter.RemoveItemFeature ( item ) ;
                  
                     if ( null != item.OwnerItem )
                     {
                        ( ( ToolStripDropDownItem ) item.OwnerItem ).DropDownItems.Remove ( item ) ;
                     }
                     else
                     {
                        item.Owner.Items.Remove ( item ) ;
                     }
                  }
               }
            
               SetActionInformation ( View, ViewerContainer, action ) ;
            }
            
            void view_SelectedActionMouseButtonChanged ( object sender, EventArgs e )
            {
               View.CanEditToolstripButtons = View.SelectedMouseButton == MedicalViewerMouseButtons.Left ;
            }

            void view_SelectedActionChanged ( object sender, EventArgs e )
            {
               SetActionInformation ( View, ViewerContainer, View.SelectedAction ) ;
            }
            
         #endregion
         
         #region Data Members
            
         #endregion
         
         #region Data Types Definition
         
            class ActionsEditorState
            {
               public ActionsEditorState ( ) 
               {
                  CustomActions           = new List<MedicalViewerActionType> ( ) ;
                  ActionAssociatedFeature = new Dictionary<MedicalViewerActionType,string> ( ) ;
               }
               
               public List <MedicalViewerActionType> CustomActions
               {
                  get ;
                  private set ;
               }
               
               public Dictionary <MedicalViewerActionType, string> ActionAssociatedFeature 
               {
                  get;
                  private set ;
               }
            }
            
         #endregion
         
      #endregion
      
      #region internal
         
         #region Methods
            
         #endregion
         
         #region Properties
            
         #endregion
         
         #region Events
            
         #endregion
         
         #region Data Types Definition
            
         #endregion
         
         #region Callbacks
            
         #endregion
         
      #endregion
   }
}

File: t/PACSFramework/VB/MedicalWorkstationCustomizationDemo/Examples/CustomizeActionsExample/ActionsEditorViewPresenter.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports Leadtools.Medical.Workstation.Presenters
Imports Leadtools.Medical.Workstation
Imports Leadtools.MedicalViewer
Imports System.Windows.Forms
Imports System.ComponentModel
Imports Leadtools.Medical.Workstation.UI

Namespace Leadtools.Demos.Workstation.Customized
   Class ActionsEditorViewPresenter
      Inherits WorkstationPresenterBase(Of IActionsEditorView)
#Region "Public"

#Region "Methods"

#End Region

#Region "Properties"

#End Region

#Region "Events"

#End Region

#Region "Data Types Definition"

#End Region

#Region "Callbacks"

#End Region

#End Region

#Region "Protected"

#Region "Methods"

      Protected Overrides Sub DoInitialize(ByVal container As WorkstationContainer, ByVal view As IActionsEditorView)
         Dim actions As MedicalViewerActionType()


         If container.State.DataServices.IsRegistered(Of ActionsEditorState)() Then
            __ActionsEditorState = container.State.DataServices.[Get](Of ActionsEditorState)()
         Else
            __ActionsEditorState = New ActionsEditorState()

            container.State.DataServices.Register(Of ActionsEditorState)(__ActionsEditorState)
         End If

         actions = GetAllowedActions(container)

         view.SetActions(actions)
         view.SetActionMouseButtons(System.Enum.GetValues(GetType(MedicalViewerMouseButtons)).OfType(Of MedicalViewerMouseButtons)().ToArray())


         If actions.Length > 0 Then
            view.SelectedAction = actions(0)

            SetActionInformation(view, container, actions(0))

            RegisterViewEvents(view)
         Else
            view.CanAddAction = False
            view.CanRemoveAction = False
         End If

         view.ActivateView(container.State.ActiveWorkstation)
      End Sub

#End Region

#Region "Properties"

#End Region

#Region "Events"

#End Region

#Region "Data Members"

#End Region

#Region "Data Types Definition"

#End Region

#End Region

#Region "Private"

#Region "Methods"

      Private Function GetAllowedActions(ByVal container As WorkstationContainer) As MedicalViewerActionType()
         Dim actions As List(Of MedicalViewerActionType)


         actions = New List(Of MedicalViewerActionType)()

         For Each action As MedicalViewerActionType In System.Enum.GetValues(GetType(MedicalViewerActionType)).OfType(Of MedicalViewerActionType)().ToArray()
            If action = MedicalViewerActionType.None OrElse action = MedicalViewerActionType.Rotate3DCamera OrElse action = MedicalViewerActionType.Rotate3DObject OrElse action = MedicalViewerActionType.RotatePlane OrElse action = MedicalViewerActionType.Scale3DObject OrElse action = MedicalViewerActionType.Translate3DCamera OrElse action = MedicalViewerActionType.TranslatePlane Then
               Continue For
            End If

            If container.State.MedicalViewerCellActions.Contains(action) Then
               If __ActionsEditorState.CustomActions.Contains(action) Then
                  actions.Add(action)
               End If
            Else
               actions.Add(action)
            End If
         Next

         Return actions.ToArray()
      End Function


      Private Sub SetActionInformation(ByVal view__1 As IActionsEditorView, ByVal container As WorkstationContainer, ByVal action As MedicalViewerActionType)
         Dim isNewAction As Boolean


         isNewAction = Not __ActionsEditorState.CustomActions.Contains(action)

         If isNewAction Then
            view__1.CanAddAction = True
            view__1.CanRemoveAction = False

            view__1.ActionDisplayName = action.ToString()
            view__1.SelectedMouseButton = MedicalViewerMouseButtons.Left

            view__1.CanEditToolstripButtons = True

            view__1.FeatureId = action.ToString() & "FeatureId"

            view__1.ToolStipItemImage = Nothing
            View.ToolStipItemAlternativeImage = Nothing
         Else
            view__1.CanAddAction = False
            view__1.CanRemoveAction = True

            Dim keyValuePair As KeyValuePair(Of MedicalViewerMouseButtons, MedicalViewerActionType) = container.State.MedicalViewerCellMouseButtonActions.Where(Function(n) n.Value = action).FirstOrDefault()

            view__1.ActionDisplayName = container.State.WorkstationActionDisplayName(action)
            view__1.SelectedMouseButton = keyValuePair.Key

            view__1.CanEditToolstripButtons = keyValuePair.Key = MedicalViewerMouseButtons.Left

            If __ActionsEditorState.ActionAssociatedFeature.ContainsKey(action) Then
               view__1.FeatureId = __ActionsEditorState.ActionAssociatedFeature(action)

               If Not String.IsNullOrEmpty(view__1.FeatureId) Then
                  Dim items As ToolStripItem() = container.StripItemFeatureExecuter.GetItems(view__1.FeatureId)

                  If items.Length > 0 AndAlso TypeOf items(0) Is IToolStripItemItemProperties Then
                     view__1.ToolStipItemImage = CType(items(0), IToolStripItemItemProperties).ItemProperties.DefaultImage
                     view__1.ToolStipItemAlternativeImage = CType(items(0), IToolStripItemItemProperties).ItemProperties.AlternativeImage
                  End If
               End If
            End If
         End If

         view__1.CanChangeDisplayName = view__1.CanAddAction
         view__1.CanChangeFeatureId = view__1.CanAddAction
         view__1.CanChangeMouseButton = view__1.CanAddAction
      End Sub

      Private Sub RegisterViewEvents(ByVal view As IActionsEditorView)
         AddHandler view.SelectedActionChanged, AddressOf view_SelectedActionChanged
         AddHandler view.SelectedActionMouseButtonChanged, AddressOf view_SelectedActionMouseButtonChanged
         AddHandler view.AddSelectedAction, AddressOf view_AddSelectedAction
         AddHandler view.RemoveSelectedAction, AddressOf view_RemoveSelectedAction
         AddHandler view.ValidateFeatureId, AddressOf view_ValidateFeatureId
      End Sub

      Private Sub view_ValidateFeatureId(ByVal sender As Object, ByVal e As CancelEventArgs)
         e.Cancel = False
         View.SetValidationErrorText(String.Empty)

         If String.IsNullOrEmpty(View.FeatureId) Then
            e.Cancel = True

            View.SetValidationErrorText("Feature ID can't be empty")
         Else
            If ViewerContainer.FeaturesFactory.IsCommandRegistered(View.FeatureId) Then
               e.Cancel = True

               View.SetValidationErrorText("Feature ID already registered")
            End If
         End If
      End Sub

      Private Sub AddCustomMenuItem(ByVal action As MedicalViewerActionType, ByVal newMenuProperties As WorkstationMenuProperties, ByVal item As ToolStripDropDownItem)
         Dim newActionItem As CustomToolStripMenuItem



         newActionItem = New CustomToolStripMenuItem(newMenuProperties)


         item.DropDownItems.Insert(0, newActionItem)

         ViewerContainer.State.ActiveWorkstation.AddToolbarSplitButtonChildItemLeftAction(CType(item, CustomToolStripSplitButton), newActionItem, action)

         ViewerContainer.StripItemFeatureExecuter.SetItemFeature(newActionItem)
      End Sub

      Private Sub AddCustomButton(ByVal action As MedicalViewerActionType, ByVal newMenuProperties As WorkstationMenuProperties, ByVal item As ToolStripDropDownItem)
         Dim newActionItem As CustomToolStripButton


         newActionItem = New CustomToolStripButton(newMenuProperties)

         item.DropDownItems.Add(newActionItem)

         ViewerContainer.State.ActiveWorkstation.AddToolbarActivatedButtonLeftAction(item, action)

         ViewerContainer.StripItemFeatureExecuter.SetItemFeature(newActionItem)
      End Sub

      Private Sub AddDesignerItem(ByVal action As MedicalViewerActionType, ByVal newMenuProperties As WorkstationMenuProperties, ByVal item As ToolStripDropDownItem)
         Dim newActionItem As DesignToolStripMenuItem


         newActionItem = New DesignToolStripMenuItem(newMenuProperties)

         item.DropDownItems.Add(newActionItem)

         ViewerContainer.StripItemFeatureExecuter.SetItemFeature(newActionItem)
      End Sub

      Private Sub AddActionToCurrentCells(ByVal action As MedicalViewerActionType)
         For Each studyViewer As StudiesViewer In ViewerContainer.State.WorkstationHostViewers.OfType(Of StudiesViewer)()
            For Each medicalViewer As MedicalViewer.MedicalViewer In studyViewer.Viewers
               For Each cell As MedicalViewerCell In medicalViewer.Cells
                  cell.AddAction(action)
               Next
            Next
         Next
      End Sub

      Private Function GetAnnotationOrRegionsParnetItems(ByVal action As MedicalViewerActionType) As ToolStripItem()
         If action.ToString().Contains("Annotation") Then
            Return ToolStripMenuProperties.Instance.AnnotationstoolStripSplitButton.AssociatedItems.OfType(Of ToolStripDropDownItem)().ToArray()
         ElseIf action.ToString().Contains("Region") Then
            Return ToolStripMenuProperties.Instance.RegionstoolStripSplitButton.AssociatedItems.OfType(Of ToolStripDropDownItem)().ToArray()
         Else
            Return Nothing
         End If
      End Function

#End Region

#Region "Properties"

      Private __ActionsEditorState As ActionsEditorState
#End Region

#Region "Events"

      Private Sub view_AddSelectedAction(ByVal sender As Object, ByVal e As EventArgs)
         Dim action As MedicalViewerActionType = View.SelectedAction
         Dim actionCommand As ApplyActionCommand

         If Not ViewerContainer.State.MedicalViewerCellActions.Contains(action) Then
            AddActionToCurrentCells(action)

            ViewerContainer.State.MedicalViewerCellActions.Add(action)
         End If

         ViewerContainer.State.WorkstationActionDisplayName(action) = View.ActionDisplayName

         __ActionsEditorState.CustomActions.Add(action)

         actionCommand = New ApplyActionCommand(View.FeatureId, ViewerContainer, action, View.SelectedMouseButton, ApplyActionFlags.Viewer2D)

         If View.SelectedMouseButton = MedicalViewerMouseButtons.Left Then
            Dim newMenuProperties As WorkstationMenuProperties


            newMenuProperties = New WorkstationMenuProperties(ViewerContainer.State.WorkstationActionDisplayName(action), View.ToolStipItemImage, View.ToolStipItemAlternativeImage)
            newMenuProperties.FeatureId = View.FeatureId

            __ActionsEditorState.ActionAssociatedFeature(action) = View.FeatureId

            ViewerContainer.FeaturesFactory.RegisterCommand(actionCommand)

            Dim parentItems As ToolStripItem() = GetAnnotationOrRegionsParnetItems(action)

            If parentItems IsNot Nothing Then
               For Each item As ToolStripDropDownItem In parentItems
                  If ViewerContainer.State.ActiveWorkstation.IsToolbarActivatedButtonLeftAction(item) Then
                     If TypeOf item Is CustomToolStripSplitButton Then
                        AddCustomMenuItem(action, newMenuProperties, item)
                     Else
                        AddCustomButton(action, newMenuProperties, item)
                     End If
                  Else
                     AddDesignerItem(action, newMenuProperties, item)
                  End If
               Next
            Else
               Dim annotationToolbarItem As CustomToolStripSplitButton = Nothing
               Dim newActionItem As CustomToolStripButton


               newActionItem = New CustomToolStripButton(newMenuProperties)

               For Each annotationItem As CustomToolStripSplitButton In ToolStripMenuProperties.Instance.AnnotationstoolStripSplitButton.AssociatedItems.OfType(Of CustomToolStripSplitButton)()
                  If annotationItem.Owner Is ViewerContainer.State.ActiveWorkstation.ViewerToolbar Then
                     annotationToolbarItem = annotationItem
                     Exit For
                  End If
               Next

               If annotationToolbarItem IsNot Nothing Then
                  ViewerContainer.State.ActiveWorkstation.ViewerToolbar.Items.Insert(ViewerContainer.State.ActiveWorkstation.ViewerToolbar.Items.IndexOf(annotationToolbarItem) - 1, newActionItem)
               Else
                  ViewerContainer.State.ActiveWorkstation.ViewerToolbar.Items.Add(newActionItem)
               End If

               ViewerContainer.State.ActiveWorkstation.AddToolbarActivatedButtonLeftAction(newActionItem, action)

               ViewerContainer.StripItemFeatureExecuter.SetItemFeature(newActionItem)
            End If
         End If

         If actionCommand.CanExecute() Then
            actionCommand.Execute()
         End If

         SetActionInformation(View, ViewerContainer, View.SelectedAction)

      End Sub

      Private Sub view_RemoveSelectedAction(ByVal sender As Object, ByVal e As EventArgs)
         Dim action As MedicalViewerActionType = View.SelectedAction


         If __ActionsEditorState.CustomActions.Contains(action) Then
            If ViewerContainer.State.MedicalViewerCellActions.Contains(action) Then
               ViewerContainer.State.MedicalViewerCellActions.Remove(action)
            End If

            ViewerContainer.State.WorkstationActionDisplayName.Remove(action)

            __ActionsEditorState.CustomActions.Remove(action)
         End If

         If __ActionsEditorState.ActionAssociatedFeature.ContainsKey(action) Then
            Dim items As ToolStripItem()


            items = ViewerContainer.StripItemFeatureExecuter.GetItems(__ActionsEditorState.ActionAssociatedFeature(action))

            For Each item As ToolStripItem In items
               If TypeOf item Is ActivatedButton AndAlso ViewerContainer.State.ActiveWorkstation.IsToolbarActivatedButtonLeftAction(item) Then
                  ViewerContainer.State.ActiveWorkstation.RemoveToolbarActivatedButtonLeftAction(item, action)
               ElseIf TypeOf item.OwnerItem Is ActivatedButton Then
                  ViewerContainer.State.ActiveWorkstation.RemoveToolbarActivatedButtonLeftAction(item.OwnerItem, action)
               End If

               ViewerContainer.StripItemFeatureExecuter.RemoveItemFeature(item)

               If item.OwnerItem IsNot Nothing Then
                  CType(item.OwnerItem, ToolStripDropDownItem).DropDownItems.Remove(item)
               Else
                  item.Owner.Items.Remove(item)
               End If
            Next
         End If

         SetActionInformation(View, ViewerContainer, action)
      End Sub

      Private Sub view_SelectedActionMouseButtonChanged(ByVal sender As Object, ByVal e As EventArgs)
         View.CanEditToolstripButtons = View.SelectedMouseButton = MedicalViewerMouseButtons.Left
      End Sub

      Private Sub view_SelectedActionChanged(ByVal sender As Object, ByVal e As EventArgs)
         SetActionInformation(View, ViewerContainer, View.SelectedAction)
      End Sub

#End Region

#Region "Data Members"

#End Region

#Region "Data Types Definition"

      Private Class ActionsEditorState
         Public Sub New()
            CustomActions = New List(Of MedicalViewerActionType)()
            ActionAssociatedFeature = New Dictionary(Of MedicalViewerActionType, String)()
         End Sub

         Public CustomActions As List(Of MedicalViewerActionType)
         Public ActionAssociatedFeature As Dictionary(Of MedicalViewerActionType, String)
      End Class

#End Region

#End Region

#Region "internal"

#Region "Methods"

#End Region

#Region "Properties"

#End Region

#Region "Events"

#End Region

#Region "Data Types Definition"

#End Region

#Region "Callbacks"

#End Region

#End Region
   End Class
End Namespace
